server/src/storage/uncompressed_data_manager.rs

```diff
@@ -0,0 +1,345 @@
+    pub fn insert_message(&mut self, message: Message) -> Result<(), String> {
+        let data_point = DataPoint::from_message(&message)?;
+        let key = data_point.generate_unique_key();
+        let _span = info_span!("insert_message", key = key.clone()).entered();
+
+        info!("Inserting data point '{}' into segment.", data_point);
+
+        if let Some(segment) = self.uncompressed_data.get_mut(&key) {
+            info!("Found existing segment.");
+
+            segment.insert_data(&data_point);
+
+            if segment.is_full() {
+                info!("Segment is full, moving it to the queue of finished segments.");
+
+                // Since this is only reachable if the segment exists in the HashMap, unwrap is safe to use.
+                let full_segment = self.uncompressed_data.remove(&key).unwrap();
+                self.enqueue_segment(key, full_segment)
+            }
+        } else {
+            info!("Could not find segment. Creating segment.");
+
+            // If there is not enough memory for a new segment, spill a finished segment.
+            if SegmentBuilder::get_memory_size() > self.uncompressed_remaining_memory_in_bytes {
+                self.spill_finished_segment();
+            }
+
+            // Create a new segment and reduce the remaining amount of reserved memory by its size.
+            let mut segment = SegmentBuilder::new();
+            self.uncompressed_remaining_memory_in_bytes -= SegmentBuilder::get_memory_size();
+
+            info!(
+                "Created segment. Remaining reserved bytes: {}.",
+                self.uncompressed_remaining_memory_in_bytes
+            );
+
+            segment.insert_data(&data_point);
+            self.uncompressed_data.insert(key, segment);
+        }
+
+        Ok(())
+    }
+
+    /// Remove the oldest [`FinishedSegment`] from the queue and return it. Return [`None`] if the
+    /// queue of finished segments is empty.
+    pub fn get_finished_segment(&mut self) -> Option<FinishedSegment> {
+        if let Some(finished_segment) = self.finished_queue.pop_front() {
+            // Add the memory size of the removed FinishedSegment back to the remaining bytes.
+            self.uncompressed_remaining_memory_in_bytes +=
+                finished_segment.uncompressed_segment.get_memory_size();
+
+            Some(finished_segment)
+        } else {
+            None
+        }
+    }
+
+    /// Move `segment_builder` to the queue of finished segments.
+    fn enqueue_segment(&mut self, key: String, segment_builder: SegmentBuilder) {
+        let finished_segment = FinishedSegment {
+            key,
+            uncompressed_segment: Box::new(segment_builder),
+        };
+
+        self.finished_queue.push_back(finished_segment);
+    }
+
+    /// Spill the first in-memory [`FinishedSegment`] in the queue of finished segments. If no
+    /// in-memory finished segments could be found, [`panic`](core::panic).
```

Comment:
As we also use [`std`](https://doc.rust-lang.org/std/index.html) and not just [`core`](https://doc.rust-lang.org/core/index.html) I suggest we link to `std` instead of `core`, e.g., [`std::panic`](https://doc.rust-lang.org/std/panic/index.html) instead of [`core::panic`](https://doc.rust-lang.org/core/panic/index.html).

Reply:
Fixed.